Reliable encoder wheel action when reversing direction

marknn3 opened the pull request:
### Description

This change will fix a reliability issue with the encoder wheel when reversing direction.
After a direction change the first step was randomly ignored.
This was caused by missed encoder inputs. In most situations this is unnoticed. But when reversing direction, this can lead to not registering the first step.

By detecting a direction change under strict conditions, this behavior has been corrected.

This fixes #18693

### Benefits

The user will have reliable encoder wheel action.
My personal reaction was that I thought I had faulty hardware. It turned out to be a software bug.

marknn3 committed:
When reversing the encoder direction, a movement step can be missed.
This happens when ABS(encoderDiff) has a non-zero residual value.
The fix will treat this condition as a full step.
Fixes #18693

GMagician commented:
It seems not to solve my issue. Again, sometimes, selection doesn't change when encoder moves...

marknn3 commented:
This patch only handles direction reversal, and really fixes a bug where a direction change does not reach the threshold where is actually should.
When you see missing steps in same direction then the software really misses (sub-)steps, or your hardware is flaky.
Note that the software must see 4 exact patterns in the right order before a single step is registered. Any hardware or timing issue will cause missing steps.

thinkyhead commented:
If you have the encoder set to do a "tick" on every 2 (or more) encoder ticks (at the hardware level) then there will be sometimes be a gap where the encoder only sends one hardware tick. I have also noticed (with the character display) that there is an extra screen update between the time you turn the encoder and when the selection has changed to the next/previous menu item. So, my longer-term plan is to dig into the encoder code and split it out as much as possible from the LCD update and click-handling code. I want to make sure all events are processed and ready before the screen handler gets any encoder data.

thinkyhead committed:
Update ultralcd.cpp

thinkyhead committed:
Update ultralcd.cpp

thinkyhead committed:
Update ultralcd.cpp

thinkyhead committed:
better

thinkyhead commented:
One trick that could work would be to only respect single encoder pulses that are very close together in time, when using a controller that sends multiple pulses per "feels like a step." I guess there are a lot of different types of encoders, and some may have smooth motion, so this is a hard thing to gauge. A way to accomplish this would be to make any residual value less than `ENCODER_PULSES_PER_STEP` expire after a short period of time, so that `encoderDiff` is never left with a residual value for more than the duration of the fastest reasonable step (about 1/10th of a second).

thinkyhead commented:
And, yet another trick that could work as a proxy for residual tick expiration is to simply reset `encoderDiff` to 0 always, and not just when `encoderPastThreshold` is set.

thinkyhead committed:
Update ultralcd.cpp

thinkyhead commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -936,7 +950,7 @@ void MarlinUI::update() {
 
           encoderPosition += (encoderDiff * encoderMultiplier) / (ENCODER_PULSES_PER_STEP);
-          encoderDiff = 0;
+          encoderDiff = 0;  // Move this and clear always?
```
Check to see if moving this down to line 964 (after the `if (encoderPastThreshold || lcd_clicked)` condition) helps to fix the issue, while also not causing the encoder to miss steps. The lcd update loop is called either 10 or 20 times per second, so that should be good enough not to miss 99.9% of multiple-pulses.

Of course, test this with the other fix disabled.

marknn3 commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -896,7 +896,21 @@ void MarlinUI::update() {
         RESET_STATUS_TIMEOUT();
 
-      const float abs_diff = ABS(encoderDiff);
+      float abs_diff = ABS(encoderDiff);
```
abs_diff should not be a float.
The only place where abs_diff is used as float is in calculating encoderMovementSteps. But even there it is questionable.
Note that the M3 has no floating point unit.

marknn3 committed:
Fix the logic, and include a check for 'past half threshold'.
This will compensate for occasional missed steps in all cases.

marknn3 committed:
Removed comment because it is not relevant anymore

GMagician commented:
Now it seems better

marknn3 commented:
I am not satisfied, it can be improved still! There are some edge-cases during/after a direction change which I am tackling now

marknn3 committed:
Different approach. The edge-cases are now fixed.
Not the cleanest implementation, unfortunately. Be VERY careful if somone wants to refactor this!
I did not see any missed or double steps so far. Very smooth operation.
This is I think the furthest we can push it for now for correcting direction changes and missed steps.

marknn3 commented:
I changed the implementation for the direction change and missing step detection, because the previous version was fundamentally wrong.
The new version works very well (for me), with smooth and reliable operation.
Unfortunately it is not the most elegant implementation. Be very careful when you try to refactor this code.

thinkyhead committed:
Tweaks while reviewing

thinkyhead commented:
> (encoderDiff > 0) == (lastEncoderDiff < 0) is not equivalent to (encoderDiff > 0 && lastEncoderDiff < 0) || (encoderDiff < 0 && lastEncoderDiff > 0)

Indeed. It misses the case where either one is equal to zero. But, I thought that case was already eliminated by the time we arrive at that condition….

thinkyhead committed:
Update ultralcd.cpp

thinkyhead committed:
Update ultralcd.cpp

thinkyhead commented:
Ok, did some minor refactoring and tried not to eat my own brain this time…
- Do the invariant update of `lastEncoderDiff` up front, grabbing the existing value first.
- Use `abs_diff` for the non-zero test and wrap in `WITHIN()`.
- Move `||` and `&&` to the front of lines.

It should function identically to your last commit. I'll install this on some test boxes while I also work on the "password" feature.

marknn3 commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -902,20 +902,21 @@ void MarlinUI::update() {
-          lastEncoderDiff = encoderDiff;
+        if (WITHIN(abs_diff, 1, (ENCODER_PULSES_PER_STEP) - 1)  // Not past threshold
+          && (abs_diff > (ENCODER_PULSES_PER_STEP) / 2          // Passed half the threshold?
+            || (ABS(encoderDiff - prevDiff) >= (ENCODER_PULSES_PER_STEP)  // Or a large change...
```
This line is not a test for a 'large change', but rather to detect a direction change. This is the actual bugfix! It replaces the '<' and '>' tests, which were not correct in all cases.

thinkyhead replied:
This individual test is non-specific, even though in context it means "reverse." In fact, it would also catch the case of a sudden large change without a reverse. But, the other conditions are what makes it (perhaps) more specific.

marknn3 replied:
Right, but this condition can only become true when stepping through zero, and thus changing direction.
This is because in all other situations (ABS(encoderDiff - prevDiff) >= (ENCODER_PULSES_PER_STEP) && ABS(prevDiff) < (ENCODER_PULSES_PER_STEP)) cannot become true here. This is not obvious.

thinkyhead replied:
I'm fine to label it as "definitely a reverse" but because it is non-obvious I want to also include a comment to point out why it _must_ be a reverse, and why it can't be a sudden large difference due to a very fast spin of the encoder.

thinkyhead commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -896,7 +896,30 @@ void MarlinUI::update() {
+                && ABS(prevDiff) < (ENCODER_PULSES_PER_STEP)    // ...starting from a partial or no step?
+               )
+             )
+        ) {
```
For different `ENCODER_PULSES_PER_STEP` values it seems good…
```cpp
if (WITHIN(abs_diff, 1, 1) && (abs_diff > 1 || (...))) // (2) Only "..." matters
```
```cpp
if (WITHIN(abs_diff, 1, 2) && (abs_diff > 1 || (...))) // (3)
```
```cpp
if (WITHIN(abs_diff, 1, 3) && (abs_diff > 2 || (...))) // (4)
```

thinkyhead committed:
Expand comments

thinkyhead commented:
Righty-o. It is more obvious that it has to be reverse, since the difference between small values only comes out to a big value when they have different signs. I'm pleased with this patch, especially since it clearly states that you can't trust hardware worth a damn.

marknn3 commented:
@thinkyhead Can you please reopen the pull request?
I have a better and much more elegant solution. A proper timeout detection eliminated the complicated tests.
It will now reset the residual value after a timeout. This was actually the starting point of this whole PR but was unable to find a proper and simple solution.
The encoder wheel in the previous version still felt a bit 'trigger happy'. The new version feels solid.
N.B. I struggle a bit with my last commit which has an unresolved merge conflict in it. This is my very first PR ever...

thinkyhead commented:
> Can you please reopen the pull request?

It's merged, so no. You be trippin'.

thinkyhead commented:
> I have a better and much more elegant solution. A proper timeout detection eliminated the complicated tests.

I look forward to reviewing it. Simple is good!
